---
title: "Using Ballgown R package for differential expression analysis in Atmosphere"
author: "Kapeel Chougule"
email: "kchougul@cshl.edu"
date: "7/4/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
Ballgown is a software package designed to facilitate flexible differential expression analysis of RNA-Seq data. It also provides functions to organize, visualize, and analyze the expression measurements for your transcriptome assembly.

We will use the output from the StringTie step that created the ballgown input files.Here we are doing a pairwise comparison for differential expression in sensitive genotype(IS20351) under Drought Stress(DS) and Well Watered(WW) condition. We have 3 replicates under each condition

> Follow the steps; copy paste the command on the R studio console and hit enter to run the code

## Ballgown analysis steps

1. set the working directory
```{r}
setwd("/Users/kchougul/data/testdata/Ballgown/tutorial_example")
```


2. install ballgown R package along with its dependencies
```{r,results='hide', message=FALSE, warning=FALSE, results='hide'}
source("http://bioconductor.org/biocLite.R")
biocLite("ballgown")
library(ballgown)
library(ggplot2)
library(gplots)
library(genefilter)
library(GenomicRanges)
library(plyr)
```


3. Read the design_matrix file
```{r}
pheno_data = read.table(file ="design_matrix", header = TRUE, sep = "\t")
```


4. full path to the sample directories
```{r}
sample_full_path <- paste("ballgown_input_files",pheno_data[,1], sep = '/')
```


5. Load ballgown data structure and save it to a variable "bg"
```{r, message=FALSE, warning=FALSE, results='hide'}
bg = ballgown(samples=as.vector(sample_full_path),pData=pheno_data)
```


6. Display a description of this object
```{r}
bg
```


7. Filter low-abundance genes. Here we remove all transcripts with a variance across the samples of less than one
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
```


8. Load gene names for lookup later in the tutorial
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```


9. Pull the gene_expression data frame from the ballgown object
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
```


10. View the first five rows of data (all columns) in one of the dataframes created
```{r}
head(gene_expression)
```


11. View the column names; change the column names to your sample names
```{r}
colnames(gene_expression) <- c("IS20351_DS_1_1","IS20351_DS_2_1","IS20351_DS_3_1","IS20351_WW_1_1","IS20351_WW_2_1","IS20351_WW_3_1")
```


12. View the row names
```{r, results='hide'}
row.names(gene_expression)
```


13. Determine the dimensions of the dataframe.  'dim()' will return the number of rows and columns
```{r}
dim(gene_expression)
```


14. Assign colors to each.  You can specify color by RGB, Hex code, or name
To get a list of color names:
```{r}
data_colors=c("tomato1","tomato2","tomato3","wheat1","wheat2","wheat3")
```


15. View expression values for the transcripts of a particular gene e.g "MSTRG.27571", then display only those rows of the data.frame
```{r}
i = row.names(gene_expression) == "MSTRG.27571"
gene_expression[i,]
```


16. What if we want to view values for a list of genes of interest all at once? 
 e,g: "MSTRG.28956" "MSTRG.28959" "MSTRG.2896"  "MSTRG.28962"
```{r}
genes_of_interest = c("MSTRG.28956", "MSTRG.28959", "MSTRG.2896", "MSTRG.28962")
i = which(row.names(gene_expression) %in% genes_of_interest)
gene_expression[i,]
```


17. Load the transcript to gene index from the ballgown object
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
```


18. Each row of data represents a transcript. Many of these transcripts represent the same gene. Determine the numbers of transcripts and unique genes  
```{r}
length(row.names(transcript_gene_table)) #Transcript count
length(unique(transcript_gene_table[,"g_id"])) #Unique Gene count
```


19. Plot #1 - the number of transcripts per gene.  
Many genes will have only 1 transcript, some genes will have several transcripts
Use the 'table()' command to count the number of times each gene symbol occurs (i.e. the # of transcripts that have each gene symbol)
Then use the 'hist' command to create a histogram of these counts
How many genes have 1 transcript?  More than one transcript?  What is the maximum number of transcripts for a single gene?
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene", main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one), paste("Genes with more than one transcript =", c_more_than_one), paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

 
20. Plot #2 - the distribution of transcript sizes as a histogram
In this analysis we supplied StringTie with transcript models so the lengths will be those of known transcripts
However, if we had used a de novo transcript discovery mode, this step would give us some idea of how well transcripts were being assembled
If we had a low coverage library, or other problems, we might get short 'transcripts' that are actually only pieces of real transcripts
```{r}
full_table <- texpr(bg , 'all')
hist(full_table$length, breaks=50, xlab="Transcript length (bp)", main="Distribution of transcript lengths", col="steelblue")
```


21. Summarize FPKM values for all samples
What are the minimum and maximum FPKM values for a particular library?
```{r}
min(gene_expression[,"IS20351_DS_1_1"])
max(gene_expression[,"IS20351_DS_2_1"])
min(gene_expression[,"IS20351_DS_3_1"])
max(gene_expression[,"IS20351_WW_1_1"])
min(gene_expression[,"IS20351_WW_2_1"])
max(gene_expression[,"IS20351_WW_3_1"])
```


22. Set the minimum non-zero FPKM values for use later.
Do this by grabbing a copy of all data values, coverting 0's to NA, and calculating the minimum or all non NA values
```{r}
min_nonzero=1
```


23. Set the columns for finding FPKM and create shorter names for figures
```{r}
data_columns=c(1:6)
short_names=c("sen_DS_1","sen_DS_2","sen_DS_3","sen_WW_1","sen_WW_2","sen_WW_3")
```


24. Plot #3 - View the range of values and general distribution of FPKM values for all  libraries
Create boxplots for this purpose
Display on a log2 scale and add the minimum non-zero value to avoid log2(0)
```{r}
boxplot(log2(gene_expression[,data_columns]+min_nonzero), col=data_colors, names=short_names, las=2, ylab="log2(FPKM)", main="Distribution of FPKMs for all 6 libraries")
```
Note that the bold horizontal line on each boxplot is the median


25. Plot #4 - plot a pair of replicates to assess reproducibility of technical replicates
Tranform the data by converting to log2 scale after adding an arbitrary small value to avoid log2(0)
```{r}
x = gene_expression[,"IS20351_DS_1_1"]
y = gene_expression[,"IS20351_DS_2_1"]
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="blue", cex=0.25, xlab="FPKM (IS20351_DS, Replicate 1)", ylab="FPKM (IS20351_DS, Replicate 2)", main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1)
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```


26. Plot #5 - Scatter plots with a large number of data points can be misleading ... regenerate this figure as a density scatter plot
```{r}
colors = colorRampPalette(c("white", "blue", "#007FFF", "cyan","#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))
smoothScatter(x=log2(x+min_nonzero), y=log2(y+min_nonzero), xlab="FPKM (IS20351_DS, Replicate 1)", ylab="FPKM (IS20351_DS, Replicate 2)", main="Comparison of expression values for a pair of replicates", colramp=colors, nbin=200)
```


27. Compare the correlation 'distance' between all replicates
Do we see the expected pattern for all  libraries (i.e. replicates most similar, then DS vs. WW)?
Calculate the FPKM sum for all 6 libraries
```{r}
gene_expression[,"sum"]=apply(gene_expression[,data_columns], 1, sum)
```


28. Identify the genes with a grand sum FPKM of at least 5 - we will filter out the genes with very low expression across the board
```{r}
i = which(gene_expression[,"sum"] > 5)
```


29. Calculate the correlation between all pairs of data
```{r}
r=cor(gene_expression[i,data_columns], use="pairwise.complete.obs", method="pearson")
r
```


30. Plot #8 - Convert correlation to 'distance', and use 'multi-dimensional scaling' to display the relative differences between libraries
This step calculates 2-dimensional coordinates to plot points for each library
Libraries with similar expression patterns (highly correlated to each other) should group together
What pattern do we expect to see, given the types of libraries we have (technical replicates, biologal replicates, DS/WW)?
```{r}
d=1-r
mds=cmdscale(d, k=2, eig=TRUE)
par(mfrow=c(1,1))
plot(mds$points, type="n", xlab="", ylab="", main="MDS distance plot (all non-zero genes) for all libraries", xlim=c(-0.15,0.15), ylim=c(-0.15,0.15))
points(mds$points[,1], mds$points[,2], col="grey", cex=2, pch=16)
text(mds$points[,1], mds$points[,2], short_names, col=data_colors)
```


31. Calculate the differential expression results including significance
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="condition", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
```


32. Plot #9 - View the distribution of differential expression values as a histogram
Display only those that are significant according to Ballgown
```{r}
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen", xlab="log2(Fold change) Sen_DS vs Sen_WW", main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```


33. Plot #10 - Display the grand expression values from UHR and HBR and mark those that are significantly differentially expressed
```{r}
gene_expression[,"Sen_DS"]=apply(gene_expression[,c(1:3)], 1, mean)
gene_expression[,"Sen_WW"]=apply(gene_expression[,c(3:6)], 1, mean)
x=log2(gene_expression[,"Sen_DS"]+min_nonzero)
y=log2(gene_expression[,"Sen_WW"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Sen_DS FPKM (log2)", ylab="Sen_WW FPKM (log2)", main="Sen_DS vs Sen_WW FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```


34. Write a simple table of differentially expressed transcripts to an output file
Each should be significant with a log2 fold-change >= 2
```{r}
sigpi = which(results_genes[,"pval"]<0.05)
sigp = results_genes[sigpi,]
sigde = which(abs(sigp[,"de"]) >= 2)
sig_tn_de = sigp[sigde,]
```


35. Order the output by or p-value and then break ties using fold-change
```{r}
o = order(sig_tn_de[,"qval"], -abs(sig_tn_de[,"de"]), decreasing=FALSE)
output = sig_tn_de[o,c("gene_name","id","fc","pval","qval","de")]
write.table(output, file="SigDE.txt", sep="\t", row.names=FALSE, quote=FALSE)
#View selected columns of the first 25 lines of output
output[1:25,c(1,4,5)]
```
